sklearn/feature_extraction/text.py

```diff
@@ -304,10 +305,35 @@ def _check_stop_words_consistency(self, stop_words, preprocess, tokenize):
             self._stop_words_id = id(self.stop_words)
             return 'error'
 
+    def _validate_analyzer(self):
+        # This is to check if the given custom analyzer expects file or a
+        # filename instead of data.
+        # Behavior changed in v0.21, function could be removed in v0.24
+        if not callable(self.analyzer):
+            return
+
+        import tempfile
+        with tempfile.NamedTemporaryFile() as f:
+            fname = f.name
+        # now we're sure fname doesn't exist
+
+        msg = ("Since v0.21, vectorizers pass the data to the custom analyzer "
+               "and not the file names or the file objects.")
+        try:
+            self.analyzer(fname)
+        except FileNotFoundError:
+            warnings.warn(msg, ChangedBehaviorWarning)
+        except AttributeError as e:
+            if str(e) == "'str' object has no attribute 'read'":
+                warnings.warn(msg, ChangedBehaviorWarning)
+        except Exception:
+            pass
```

Comment:
pass??? Do you mean raise?

Reply:
I actually had it as `raise` first :D But here I'm only checking if the custom analyzer raises those exceptions, and raise a warning if it does. When the code continues, it'll call the analyzer with a text and the same exception will be raised and it'll be shown to the user.

Example:

```python
from sklearn.feature_extraction.text import HashingVectorizer

def analyzer(x):
    with _ in x.read():
        pass

HashingVectorizer(analyzer=analyzer,
                  input='file').fit_transform(['hello world'])

/path/to/sklearn/feature_extraction/text.py:328: ChangedBehaviorWarning: Since v0.21, vectorizers pass the data to the custom analyzer and not the file names or the file objects.
  warnings.warn(msg, ChangedBehaviorWarning)
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
~/Projects/github.com/sklearn/bugs/countvectorizer.py in <module>
      5         pass
      6 
----> 7 HashingVectorizer(analyzer=analyzer, input='file').fit_transform(['hello world'])
      8 
      9 #cv = TfidfVectorizer(analyzer=lambda x: x.split(), input='file')

~/Projects/github.com/sklearn/scikit-learn/sklearn/feature_extraction/text.py in fit_transform(self, X, y)
    665             Document-term matrix.
    666         """
--> 667         return self.fit(X, y).transform(X)
    668 
    669     def _get_hasher(self):

~/Projects/github.com/sklearn/scikit-learn/sklearn/feature_extraction/text.py in transform(self, X)
    640 
    641         analyzer = self.build_analyzer()
--> 642         X = self._get_hasher().transform(analyzer(doc) for doc in X)
    643         if self.binary:
    644             X.data.fill(1)

~/Projects/github.com/sklearn/scikit-learn/sklearn/feature_extraction/hashing.py in transform(self, raw_X)
    152         indices, indptr, values = \
    153             _hashing_transform(raw_X, self.n_features, self.dtype,
--> 154                                self.alternate_sign)
    155         n_samples = indptr.shape[0] - 1
    156 

~/Projects/github.com/sklearn/scikit-learn/sklearn/feature_extraction/_hashing.pyx in sklearn.feature_extraction._hashing.transform()
     48     cdef np.ndarray values = np.empty(capacity, dtype=dtype)
     49 
---> 50     for x in raw_X:
     51         for f, v in x:
     52             if isinstance(v, (str, unicode)):

~/Projects/github.com/sklearn/scikit-learn/sklearn/feature_extraction/hashing.py in <genexpr>(.0)
    149             raw_X = (_iteritems(d) for d in raw_X)
    150         elif self.input_type == "string":
--> 151             raw_X = (((f, 1) for f in x) for x in raw_X)
    152         indices, indptr, values = \
    153             _hashing_transform(raw_X, self.n_features, self.dtype,

~/Projects/github.com/sklearn/scikit-learn/sklearn/feature_extraction/text.py in <genexpr>(.0)
    640 
    641         analyzer = self.build_analyzer()
--> 642         X = self._get_hasher().transform(analyzer(doc) for doc in X)
    643         if self.binary:
    644             X.data.fill(1)

~/Projects/github.com/sklearn/scikit-learn/sklearn/feature_extraction/text.py in <lambda>(doc)
    334         self._validate_analyzer()
    335         if callable(self.analyzer):
--> 336             return lambda doc: self.analyzer(self.decode(doc))
    337 
    338         preprocess = self.build_preprocessor()

~/Projects/github.com/sklearn/scikit-learn/sklearn/feature_extraction/text.py in decode(self, doc)
    135 
    136         elif self.input == 'file':
--> 137             doc = doc.read()
    138 
    139         if isinstance(doc, bytes):

AttributeError: 'str' object has no attribute 'read'
```

Reply:
Right. I see clearer now. But shouldn't this logic only be correct for input='filename'?

Reply:
I think the warning also needs to be raised if input='file' and the analyzer is passed a file which it tries to treat as a string

Reply:
Ok, validating only if input is either file or filename.